django/db/models/sql/query.py

```diff
@@ -890,18 +895,29 @@ def join(self, join, reuse=None):
         The 'reuse' parameter can be either None which means all joins are
         reusable, or it can be a set containing the aliases that can be reused.
 
+        The 'reuse_with_filtered_relation' is used when computing
+        FilteredRelation's instances.
+
         A join is always created as LOUTER if the lhs alias is LOUTER to make
         sure chains like t1 LOUTER t2 INNER t3 aren't generated. All new
         joins are created as LOUTER if the join is nullable.
         """
-        reuse = [a for a, j in self.alias_map.items()
-                 if (reuse is None or a in reuse) and j == join]
-        if reuse:
-            self.ref_alias(reuse[0])
-            return reuse[0]
+        if reuse_with_filtered_relation and reuse:
+            reuse_aliases = [
+                a for a, j in self.alias_map.items()
+                if a in reuse and j.equals(join, with_filtered_relation=False)
+            ]
+        else:
+            reuse_aliases = [
+                a for a, j in self.alias_map.items()
+                if (reuse is None or a in reuse) and j == join
+            ]
```

Comment:
What I meant was that the code could be simplifed a bit more like...

```py
reuse_aliases = [
    a for a, j in self.alias_map.items()
    if (reuse is None or a in reuse) and
       j.equals(join, with_filtered_relation=reuse_with_filtered_relation)
]
```

However you'd have to invert the meaning of `reuse_with_filtered_relation`. I actually find it a confusing parameter as it stands, as by passing it as `True`, it passes `with_filtered_relation` as `False`. I think I was a bit mistaken when I suggested the rename.

Reply:
I don't know. I also like the approach of having only one parameter `can_reuse` carrying the data.
`reuse_with_filtered_relation` as a boolean feels more lightweight and easier to understand its meaning rather than `force_reuse`. But it is less generic, if at some point it needs to be.
I'm +0 with `reuse_with_filtered_relation`

Reply:
Ok let's keep `reuse_with_filtered_relation`. At least it's very greppable.